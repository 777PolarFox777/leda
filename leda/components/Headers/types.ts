import * as React from 'react';

export interface HeaderProps extends React.HTMLAttributes<HTMLHeadingElement> {
  ref?: HeaderRefCurrent,
  shouldRender?: boolean,
  [x: string]: unknown,
}

export interface HeaderRefCurrent {
  wrapper: HTMLHeadingElement | null,
}
