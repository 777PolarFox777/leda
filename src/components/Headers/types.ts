import * as React from 'react';

export interface HeaderProps extends React.HTMLAttributes<HTMLHeadingElement> {
  ref?: HeaderRefCurrent,
  [x: string]: unknown,
}

export interface HeaderRefCurrent {
  wrapper: HTMLHeadingElement | null,
}
