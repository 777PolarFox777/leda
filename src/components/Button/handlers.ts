import * as React from 'react';
import { isFunction } from 'lodash';
import { getForms, validate } from '../Validation';
import { ButtonProps } from './types';
import { fromFormArraytoFormObject } from './helpers';

export const createClickHandler = (props: ButtonProps) => (ev: React.MouseEvent<HTMLButtonElement>): void => {
  const {
    onClick, onValidationFail, isDisabled, isLoading,
    scrollOffset, form: formProp, shouldScrollToInvalidFields,
  } = props;

  if (isDisabled || isLoading) return; // если кнопка отключена или в состояниии загрузки, прервать выполнение функции

  // если кнопка участвует в валидации форм
  if (formProp) {
    const buttonFormNames = Array.isArray(formProp) ? formProp : [formProp];

    const isEachFormValid = buttonFormNames
      .map(currentForm => validate(currentForm)) // validate all forms
      .every(item => item); // tell me if all of them are valid

    if (!isEachFormValid) {
      const invalidForms = getForms()
        .filter(currentForm => buttonFormNames.includes(currentForm.name))
        .filter(currentForm => currentForm.fields.some(field => !field.isValid));

      if (isFunction(onValidationFail)) {
        onValidationFail({ ...ev, invalidForms });
      }

      if (shouldScrollToInvalidFields) {
        const firstInvalidFormName = invalidForms[0].name;
        // data-form для buttonGroup, использовать просто [from=""] нельзя, т.к. захватит кнопки тоже
        const formElements = document.querySelectorAll(`input[form="${firstInvalidFormName}"], [data-form="${firstInvalidFormName}"`);
        // ждем пока обновятся данные
        setTimeout(() => {
          const invalidElement = Array.from(formElements).find(element => element.getAttribute('aria-invalid') === 'true');

          if (invalidElement) {
            const invalidElementRect = invalidElement.getBoundingClientRect();
            const isIE = !!(document as any).documentMode || /Edge/.test(navigator.userAgent);
            const offset = invalidElementRect.top - (scrollOffset ?? 0);
            if (isIE) {
              window.scrollBy(0, offset);
            } else {
              window.scrollBy({
                top: offset,
                behavior: 'smooth',
              });
            }
          }
        }, 0);
      }

      return;
    }
  }

  if (isFunction(onClick)) {
    if (!formProp) {
      onClick(ev);
      return;
    }

    const forms = getForms(formProp);
    const formsObject = fromFormArraytoFormObject(forms);

    const customEvent = {
      ...ev,
      forms,
      form: formsObject,
    };

    onClick(customEvent);
  }
};
