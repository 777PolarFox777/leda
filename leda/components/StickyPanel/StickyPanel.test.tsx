import React from 'react';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';
import { StickyPanel } from './index';
import { Button } from '../Button';
import { H1 } from '../Headers';

describe('StickyPanel SNAPSHOTS', () => {
  it('should render', () => {
    const panel = (
      <StickyPanel>
        <Button _success>OK</Button>
        <Button>Cancel</Button>
      </StickyPanel>
    );

    const wrapper = mount(panel);

    expect(wrapper.find('div.stickypanel-wrapper')).toHaveLength(1);

    expect(wrapper.find('.stickypanel-container')).toHaveLength(1);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('StickyPanel ATTRIBUTES', () => {
  it('should have children', () => {
    const panel = (
      <StickyPanel>
        <H1>Would you like to continue?</H1>
        <Button success>OK</Button>
        <Button>Cancel</Button>
      </StickyPanel>
    );

    const wrapper = mount(panel);

    expect(wrapper.containsAllMatchingElements([
      <H1>Would you like to continue?</H1>,
      <Button success>OK</Button>,
      <Button>Cancel</Button>,
    ])).toBeTruthy();
  });

  it('should accept componentOffset', () => {
    const panel = (
      <StickyPanel componentOffset={200}>
        <H1>Would you like to continue?</H1>
        <Button success>OK</Button>
        <Button>Cancel</Button>
      </StickyPanel>
    );

    const wrapper = mount(panel); // eslint-disable-line
    // todo: не создает отступ, дописать тест, когда починят
  });
});
