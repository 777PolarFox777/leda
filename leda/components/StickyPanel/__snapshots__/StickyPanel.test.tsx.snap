// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StickyPanel SNAPSHOTS should render 1`] = `
<StickyPanel>
  <Div
    className="stickypanel-wrapper"
  >
    <div
      className="stickypanel-wrapper"
    >
      <div
        className="stickypanel-container"
        style={Object {}}
      >
        <Button
          _success={true}
        >
          <button
            className="button-wrapper success"
            onClick={[Function]}
            type="button"
          >
            OK
          </button>
        </Button>
        <Button>
          <button
            className="button-wrapper"
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
        </Button>
      </div>
    </div>
  </Div>
</StickyPanel>
`;
